File: Product_Of_Array_Except_Self/Product_Of_Array_Except_Self.cpp
#include <vector>
#include <iostream>

using namespace std;

class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int pre = 1;
        int post = 1;
        int size = nums.size();
        vector<int> ans(size);
        for(int i = 0; i < nums.size(); i++){
           ans[i] = pre;
           pre = nums[i]*pre;
        }
        for(int i = nums.size() - 1; i >= 0; --i){
            ans[i] = ans[i] * post;
            post = post*nums[i];
        }
        return ans; 
    }
};

int main() {
    Solution solution;
    vector<int> nums = {1, 2, 3, 4}; 
    vector<int> result = solution.productExceptSelf(nums);
    
    cout << "Output: [";
    for (size_t i = 0; i < result.size(); ++i) {
        cout << result[i];
        if (i != result.size() - 1) cout << ", ";
    }
    cout << "]" << endl;
    
    return 0;
}

File: Product_Of_Array_Except_Self/Product_Of_Array_Except_Self.py
from typing import List
from collections import defaultdict

class Solution:
    def productExceptSelf(self, nums: List[int] ) -> List[int]:
        pre = 1
        post = 1
        n = len(nums)
        ans = [1] * n
        for i in range(n):
            ans[i] = pre
            pre = nums[i]*pre
        for i in reversed(range(n)):
            ans[i] = ans[i] * post
            post = post*nums[i]
        return ans

sol = Solution()
nums = [1,2,3,4]
result = sol.productExceptSelf(nums)
print(result)
